### PR TITLE
Adding ollama support

### PR DESCRIPTION
**Pull request by CyrilPeponnet**
### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [x] Describes the source of new concepts.
- [x] References existing implementations as appropriate.
- [ ] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

Address #307

### CODE REVIEW

**Comment by tmc** [comment]
Excited to see this!

**Comment by CyrilPeponnet** [comment]
Example of use:
```go
	llm, err := ollama.New(
		ollama.WithModel("zephyr"),
	)
```

The model is required. One might still pass an alternative model with the WithModel llms.Options during a Call invocation. 

There is a bunch of options that you can set for the runner or the query itself that are not in the main llms.Options. 

Also as Ollama is templating the prompt query depending on the model you use, its disabled by default as it doesn't play well with agents and other templating done on the chain level.

**Comment by CyrilPeponnet** [comment]
Adding the chat model support and all passing it back to draft for now.

**Comment by CyrilPeponnet** [comment]
The chatModel support requires some knowledge about the the model that is running on the other side as the multiturn prompt is not handled by ollama by default. This implementation uses llama2 multi turn prompt syntax but you can provide your own if you are toying with Zephyr for instance.

I also added the embedding support even if the embedding provided by llamaccp is for now pretty limited, this might change in the future with the support of Bert models dedicated for document/query embedding.

**Comment by tmc** [comment]
Fantastic work! sorry about the delay in review here. I'm actually thinking we should remove our "local" implementation as it was never really fleshed out and would love to lean further into ollama support!

File: embeddings/ollama/ollama.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package ollama
```
Comment (tmc) [inline embeddings/ollama/ollama.go:1]: please include doc.go files in new packages with a package-level description.

File: embeddings/ollama/ollamachat/ollama_chat.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+func NewChatOllama(opts ...ChatOption) (ChatOllama, error) {
+	o, err := applyChatClientOptions(opts...)
+	if err != nil {
+		return ChatOllama{}, err
+	}
+
+	return o, nil
+}
+
+func (e ChatOllama) EmbedDocuments(ctx context.Context, texts []string) ([][]float32, error) {
```
Comment (tmc) [inline embeddings/ollama/ollamachat/ollama_chat.go:31]: can you include a comment on this?

File: embeddings/ollama/ollamachat/ollama_chat.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			return nil, err
+		}
+
+		emb = append(emb, combined)
+	}
+
+	return emb, nil
+}
+
+func (e ChatOllama) EmbedQuery(ctx context.Context, text string) ([]float32, error) {
```
Comment (tmc) [inline embeddings/ollama/ollamachat/ollama_chat.go:60]: as above

**Review by tmc — APPROVED** [review]
There are some things to spruce up in here but overall this looks really great! Merging!

**Comment by eliben** [comment]
@CyrilPeponnet can you please clarify why there are two very similar implementations of embeddings in `embeddings/ollama/ollama.go` and `embeddings/ollama/ollamachat/ollama_chat.go`? They seem essentially identical except that their `client` has a different type?

**Pull request merged**